org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/preferences/PersistentModel.java

```diff
@@ -8,34 +8,42 @@
 
 package org.eclipse.buildship.core.preferences;
 
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.base.Optional;
+
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IClasspathEntry;
+
 /**
- * Interface to load and store Gradle model elements stored in the workspace plugin state area. The
- * model elements are stored and retrieved in a map-like fashion.
+ * Interface to load and store Gradle model elements stored in the workspace plugin state area.
+ * <p/>
+ * The {@link Optional} return type on the getters mark whether a value is present in the model.
  *
  * @author Donat Csikos
  */
 public interface PersistentModel {
 
-    /**
-     * Returns a value for the specified key.
-     *
-     * @param key the key
-     * @param defaultValue the to be returned if no value is present
-     * @return the value
-     */
-    String getValue(String key, String defaultValue);
-
-    /**
-     * Stores a new key-value pair in this instance. To persist the update, the {@link #flush()}
-     * method should be called.
-     *
-     * @param key the key
-     * @param value the value
-     */
-    void setValue(String key, String value);
-
-    /**
-     * Persists all changes on the disk.
-     */
-    void flush();
+    Optional<IPath> getBuildDir();
+
+    void setBuildDir(Optional<IPath> buildDirPath);
+
+    Collection<IPath> getSubprojectPaths();
+
+    void setSubprojectPaths(Collection<IPath> subprojectPaths);
+
+    Optional<List<IClasspathEntry>> getClasspath();
+
+    void setClasspath(List<IClasspathEntry> classpath);
+
+    Collection<IResource> getDerivedResources();
+
+    void setDerivedResources(Collection<IResource> derivedResources);
```

Comment:
I think these should be IPath instances like buildDir and Subprojects. Their presence can change over time. E.g. the `.gradle` folder might not yet be there when we synchronize the project.